### PR TITLE
[Assets] Call implementation for `transfer_all`

### LINKED ISSUE #4517: `transfer_all` Call in Assets Pallet
With the ability to pay fees in assets, there is demand from some users to be able to "pay fees with and transfer all", completely killing the balance of the asset on the account. We should add this `transfer_all` call to the Assets pallet to make this easier without needing to calculate fees/exchange rates to determine transferrable amount.

@pandres95 ?

### PR DESCRIPTION
**Pull request by pandres95**
Closes #4517

Polkadot address: 12gMhxHw8QjEwLQvnqsmMVY1z5gFa54vND74aMUbhhwN6mJR

### CODE REVIEW

File: prdoc/pr_4527.prdoc
Comment (joepetrowski) [inline prdoc/pr_4527.prdoc:23]: Should this be minor? It doesn't break any interface.

File: substrate/frame/assets/src/lib.rs
Comment (joepetrowski) [inline substrate/frame/assets/src/lib.rs:1710]: Assets have a `minimum_balance` but no existential deposit. This needs to be reworded. Also does `keep_alive` refer to `Accounts` in the context of the Assets pallet (i.e. `(AssetId, AccountId)`) or in the context of the system?

Comment (pandres95) [reply substrate/frame/assets/src/lib.rs:1710]: In this case, it refers to the context of Assets pallet. I'll make sure to reword that part.

**Review by joepetrowski — APPROVED** [review]
Sorry that I missed the ping @pandres95

File: prdoc/pr_4527.prdoc
Comment (muharem) [inline prdoc/pr_4527.prdoc:17]:
```suggestion
```

File: substrate/frame/assets/src/benchmarking.rs
Comment (muharem) [inline substrate/frame/assets/src/benchmarking.rs:557]: better use min_balance of the asset id here instead, multiply it by two for example.

File: substrate/frame/assets/src/tests.rs
Comment (muharem) [inline substrate/frame/assets/src/tests.rs:826]: its merged now

File: prdoc/pr_4527.prdoc
Comment (muharem) [inline prdoc/pr_4527.prdoc:11]: new call is new enum variant within Call enum. so this is a breaking change as for a rust crate.

File: prdoc/pr_4527.prdoc
Comment (muharem) [inline prdoc/pr_4527.prdoc:23]:
```suggestion
    bump: major
```

File: prdoc/pr_4527.prdoc
Comment (joepetrowski) [reply prdoc/pr_4527.prdoc:11]: That's rough. I guess we have to follow crates rules but it's not breaking for transaction encoding.

Comment (muharem) [reply prdoc/pr_4527.prdoc]: Yes, but for crates versioning I think it should follow the rust semver rules. At least that's what we have not in guidelines.

File: substrate/frame/assets/src/benchmarking.rs
Comment (muharem) [inline substrate/frame/assets/src/benchmarking.rs:557]: I meant something like `Assets::<T, I>::minimum_balance(asset_id) * 2`

File: prdoc/pr_4527.prdoc
Comment (joepetrowski) [reply prdoc/pr_4527.prdoc]: Yeah, @ggwpez also pointed out that it breaks the `WeightInfo` trait for anyone using this pallet. So I agree it is (unfortunately) major.

**Review by muharem — APPROVED** [review]

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/2024-07-16-technical-fellowship-opendev-call/9512/1

**Comment by joepetrowski** [comment]
@pandres95 is this waiting for anything before merging?

**Comment by pandres95** [comment]
@joepetrowski not that I know. I thought it was waiting for an additional review / auditing process. 😅

File: substrate/frame/assets/src/lib.rs
Comment (gui1117) [inline substrate/frame/assets/src/lib.rs:1779]: weight is calling the wrong function.

Comment (joepetrowski) [reply substrate/frame/assets/src/lib.rs:1779]: Ahh good catch

**Pull request merged**
